test/tutorials/calculational_logic.jl

```diff
@@ -20,9 +20,9 @@ include(joinpath(dirname(pathof(Metatheory)), "../examples/calculational_logic_t
   params = SaturationParams(timeout = 12, eclasslimit = 10000, schedulerparams = (match_limit = 1000, ban_length = 5))
   @test prove(calculational_logic_theory, :(((p ⟹ (p || p)) == ((!(p) && q) ⟹ q))), 1, 10, params)
 
-  ex = :((p ⟹ (q ⟹ r)) ⟹ ((p ⟹ q) ⟹ (p ⟹ r)))   # Frege's theorem
+  freges = :((p ⟹ (q ⟹ r)) ⟹ ((p ⟹ q) ⟹ (p ⟹ r)))   # Frege's theorem
   params = SaturationParams(timeout = 12, eclasslimit = 10000, schedulerparams = (match_limit = 6000, ban_length = 5))
-  @test prove(calculational_logic_theory, ex, 2, 10, params)
+  @test_broken true == prove(calculational_logic_theory, freges, 2, 10, params)
```

Comment:
This fails now because saturation reaches the enodelimit before the theory can be proven. Increasing the enodelimit fixes the test. 

The question is why it produces more enodes....

The reason seems to be the different order of applying rules (vector instead of stack).
In fact, this test succeeds:
```julia
  @test true == prove(reverse(calculational_logic_theory), freges, 2, 10, params)
```

Reply:
I would argue that the order of applying rules should not matter for EqSat. 

I'm not 100% sure why reversing the order has an effect on the number of enodes after x iterations.

Reply:
The issue that the order of rules affects the results is probably not caused by the changes in this PR but already exists in ale/3.0.

Reply:
We should keep this.

Reply:
@gkronber could it be something related to `analysis_pending` and rebuilding is processed? 

Adding this
```
function rebuild_memo!(g::EGraph)
  new_memo = Dict{VecExpr,Id}()
  for eclass in values(g.classes)
    for node in eclass.nodes
      canonicalize!(g, node)
      new_memo[node] = eclass.id
    end
  end
  g.memo = new_memo
end
```

After every rebuild didn't seem to solve the issue at all, so it looks like the memo is alright

Reply:
I think we should not concern ourself with rebuilding in this PR. The changes in this PR are ok and we can fix the failing test by increasing the enodelimit. I have a branch ready with several improvements of the rebuild procedure.